File: Codes/Class_2/secant_method.cpp
#include <bits/stdc++.h>

using namespace std;

double f(double x) {
    return x * x - 4 * x - 10;
}

double x_next(double x, double x_prev) {
    return x - (f(x) * (x - x_prev)) / (f(x) - f(x_prev));
}

int main()
{
    cout << fixed << setprecision(5);

    double x1 = 4, x2 = 2, x3;
    double prev_x1 = x1, prev_x2 = x2;
    int i = 1;

    while (1) {
        x3 = x_next(x2, x1);

        x1 = x2;
        x2 = x3;
        
        cout << "Iteration " << i << ": " << x3 << "\n";
        i++;

        if (abs(prev_x1 - x1) < 1e-6 && abs(prev_x2 - x2) < 1e-6) {
            break;
        }

        prev_x1 = x1;
        prev_x2 = x2;
    }


    return 0;
}

File: Codes/Class_2/secant_method.py
import math

func = input("Enter the function: ")

def f(x):
    return eval(func)

def x_next(x, x_prev):
    return x - (f(x) * (x - x_prev)) / (f(x) - f(x_prev))

x1, x2 = map(float, input("Enter x1 and x2: ").split())

prev_x1 = x1
prev_x2 = x2
x3 = 0

i = 1

while True:
    x3 = x_next(x2, x1)

    x1 = x2
    x2 = x3
    
    print(f"Iteration {i} : {x3}")
    i += 1

    if abs(prev_x1 - x1) < 1e-6 and abs(prev_x2 - x2) < 1e-6:
        break

    prev_x1 = x1
    prev_x2 = x2

print(f"Root : {x3}")